lib/libc/stdlib/malloc/mrs/mrs.3

```diff
@@ -0,0 +1,231 @@
+complete.
+This results in the revocation of all pointers staged for quarantine by
+.Fn free
+or
+.Fn realloc .
+.Sh ENABLING REVOCATION
+Because heap temporal safety is an experimental feature with performance
+impacts that are not fully characterized,
+.Nm
+provides a number of policy mechanism to control enabling quarantine making
+use of
+.Xr cheri_revoke 2 .
+When a program using
+.Nm
+starts up, an array of parameters are evaulated to determine if free'd
+objects should be quarantined and revoked or if the underlying
+implementations should be exposed directly.
+MRS considers the following factors in order of incresing precidence:
+.Bl -bullet
+.It
+The value of the
+.Dv security.cheri.runtime_quarantine_default
+sysctl value.
+This is exposed by the kernel via the presence or absence of the
+.Dv ELF_BSDF_CHERI_REVOKE
+flag and the absense of the
+.Dv ELF_BSDF_CHERI_REVOKE_FORCED
+flag in the
+.Dv AT_BSDFLAGS
+ELF auxiliary argument variable.
+.It
+The value of the
+.Va malloc_quarantining
+variable if it is
+.Dv MQ_ENABLED
+or
+.Dv MQ_DISABLED .
+.Va malloc_quarantining
+is a weak symbol in libc which can be provided by the program to over
+ride the system default or prevent administrative overrides via
+.Xr proccontrol 1
+or
+.Xr elfctl 1 .
+For example, the following will enable revocation without regard to the
+system-wide default:
+.Dl int malloc_quarantining = MQ_ENABLED;
+.It
+The presence of
+.Dv NT_FREEBSD_FCTL_CHERI_REVOKE_DISABLE
+or
+.Dv NT_FREEBSD_FCTL_CHERI_REVOKE_ENABLE
+(with the latter taking precidence if both are set) in the FreeBSD feature
+control ELF note as set by
+.Xr elfctl .
+This is exposed by the kernel via the presence or absence of the
+.Dv ELF_BSDF_CHERI_REVOKE
+flag and the presence of the
+.Dv ELF_BSDF_CHERI_REVOKE_FORCED
+flag in the
+.Dv AT_BSDFLAGS
+ELF auxiliary argument variable.
+.It
+The presence of
+.Dv P2_CHERI_REVOKE_ENABLE
+or
+.Dv P2_CHERI_REVOKE_DISABLE
+(with the latter taking precidence if both are set) in the process flags
+as controlled by
+the
+.Xr procctl 2
+.Dv PROC_CHERI_REVOKE_CTL
+command which is usually set by
+.Xr proccontrol 1 .
+.It
+The value of the
+.Va malloc_quarantining
+variable if it is
+.Dv MQ_ENABLED_FORCED
+or
+.Dv MQ_DISABLED_FORCED .
+.El
+.Pp
+A program can determing if
+.Nm
+revocation is enabled by checking the return value of
+.Fn malloc_is_revoking .
+.Sh TRACING
+.Nm
+supports
+.Xr utrace 2
+event logging which can be enabled by setting the
+.Ev _MRS_UTRACE
+environmental variable, collected by
+.Xr ktrace 1
+and viewed by
+.Xr kdump 1 .
+This produces output like:
+.Pp
+.Dl $ kdump -H -t u | grep 0x45e00000
+.Dl  21593 101243 git      USER  0x45e00000 = malloc(50708)
+.Dl  21593 101243 git      USER  0x45e00000 = mrs_malloc(50708)
+.Dl  21593 101243 git      USER  mrs_free(0x45e00000)
+.Dl  21593 101243 git      USER  quarantine_insert(0x45e00000, 57344)
+.Dl  21593 101242 git      USER  free(0x45e00000)
```

Comment:
```suggestion
.Dl  21593 101242 git      USER  quarantine_flush()
```

The free() log was actually from jemalloc, not MRS, and I added a quarantine_flush() trace today while reworking the multithread stuff.